doc/NAT/SAI-NAT-API.md

```diff
@@ -0,0 +1,297 @@
+
+Similarly detect the absence of NAT object in switch and if NULL then disable the HW global register for NAT.
+
+## NAT Object Workflow
+Following is a logical representation of SAI NAT objects
+
+>Example 1: SNAT and DNAT
+From Zone ID: 100
+To Zone ID: 200
+VRF: None
+STATIC_NAT|65.55.42.1:1024 
+InternalEndpoint: 10.0.0.1:6000
+STATIC_NAT|65.55.42.1:1025
+InternalEndpoint: 10.0.0.2:6000
+STATIC_NAT|65.55.42.1:1026
+InternalEndpoint: 10.0.0.3:6000
+
+##### Step 1: Create a Source NAT Entry object:
+```sh
+# First Translation
+sai_nat_entry_attr_list[0].id = SAI_NAT_ENTRY_ATTR_FROM_IP;
+sai_nat_entry_attr_list[0].value.ipaddr.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+sai_nat_entry_attr_list[0].value.ipaddr.ip4 = 10.0.0.1; 
+
+>sai_nat_entry_attr_list[1].id = SAI_NAT_ENTRY_ATTR_TO_IP;
+sai_nat_entry_attr_list[1].value.ipaddr.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+sai_nat_entry_attr_list[1].value.ipaddr.ip4 = 65.55.42.1; 
+
+>sai_nat_entry_attr_list[2].id = SAI_NAT_ENTRY_ATTR_FROM_PORT;
+sai_nat_entry_attr_list[2].value.u16 = 6000;
+
+sai_nat_entry_attr_list[3].id = SAI_NAT_ENTRY_ATTR_TO _PORT;
+sai_nat_entry_attr_list[3].value.u16 = 1024;
+
+sai_nat_entry_attr_list[4].id = SAI_NAT_ENTRY_ATTR_IP_PROTOCOL;
+sai_nat_entry_attr_list[4].value.u8 = 17; 
+
+# Second Translation
+sai_nat_entry_attr_list[5].id = SAI_NAT_ENTRY_ATTR_FROM_IP;
+sai_nat_entry_attr_list[5].value.ipaddr.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+sai_nat_entry_attr_list[5].value.ipaddr.ip4 = 10.0.0.2; 
+
+sai_nat_entry_attr_list[6].id = SAI_NAT_ENTRY_ATTR_TO_IP;
+sai_nat_entry_attr_list[6].value.ipaddr.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+sai_nat_entry_attr_list[6].value.ipaddr.ip4 = 65.55.42.1; 
+
+sai_nat_entry_attr_list[7].id = SAI_NAT_ENTRY_ATTR_FROM_PORT;
+sai_nat_entry_attr_list[7].value.u16 = 6000;
+
+sai_nat_entry_attr_list[8].id = SAI_NAT_ENTRY_ATTR_TO _PORT;
+sai_nat_entry_attr_list[8].value.u16 = 1025;
+
+sai_nat_entry_attr_list[9].id = SAI_NAT_ENTRY_ATTR_IP_PROTOCOL;
+sai_nat_entry_attr_list[9].value.u8 = 17; 
+
+# Third Translation
+sai_nat_entry_attr_list[10].id = SAI_NAT_ENTRY_ATTR_FROM_IP;
+sai_nat_entry_attr_list[10].value.ipaddr.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+sai_nat_entry_attr_list[10].value.ipaddr.ip4 = 10.0.0.3; 
+
+sai_nat_entry_attr_list[11].id = SAI_NAT_ENTRY_ATTR_TO_IP;
+sai_nat_entry_attr_list[11].value.ipaddr.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+sai_nat_entry_attr_list[11].value.ipaddr.ip4 = 65.55.42.1; 
+
+sai_nat_entry_attr_list[12].id = SAI_NAT_ENTRY_ATTR_FROM_PORT;
+sai_nat_entry_attr_list[12].value.u16 = 6000;
+
+sai_nat_entry_attr_list[13].id = SAI_NAT_ENTRY_ATTR_TO _PORT;
+sai_nat_entry_attr_list[13].value.u16 = 1026;
+
+sai_nat_entry_attr_list[14].id = SAI_NAT_ENTRY_ATTR_IP_PROTOCOL;
+sai_nat_entry_attr_list[14].value.u8 = 17; 
+
+nat_entry_attr_count = 15;
+sai_create_nat_entry_fn(

```

Comment:
Its upto the caller. If bulking is needed and all entries have same masks then there is no need to create 3 entry objects. If the entries do not have same attributes then 3 separate objects MUST be created.

Reply:
You are providing the same attribute ID several times in single create call function with different values. I am not familiar with any other API that allows that in SAI

There is bulk create definition in SAI
typedef sai_status_t (*sai_bulk_object_create_fn)(
        _In_ sai_object_id_t switch_id,
        _In_ uint32_t object_count,
        _In_ const uint32_t *attr_count,
        _In_ const sai_attribute_t **attr_list,
        _In_ sai_bulk_op_error_mode_t mode,
        _Out_ sai_object_id_t *object_id,
        _Out_ sai_status_t *object_statuses);

If you want to allow bulk creation, I think we should have standard create and a bulk create
But using standard create as bulk create isn't something which is supported, IMO

Reply:
This is not a usual case of attribute list. Its 1:1 mapping for NAPT.
Other way we can do is create each entry as an object then have sai_object_list_t. But this will create an explosion of objects.
Do you have a suggestion?

Reply:
I think either to create each entry as an object. And I agree, that the downside will be many objects.
The other option is similar to qos maps sai_qos_map_list_t, to use a list of structs. The downside is you can't use strong parameter checking by the meta data.
I think the first option is the prefered way, and where most of the later specs have went (for example, tunnel maps which are composed of multiple entries each are an object)

Reply:
Lets discuss this in community.
Given the scale of NAT entries will >20K, 60k< , creating that many objects is prohibitively expensive.